pytorch_lightning/utilities/cli.py

```diff
@@ -165,20 +169,23 @@ def __init__(
         self.add_arguments_to_parser(self.parser)
         self.parse_arguments()
         if self.config['seed_everything'] is not None:
-            seed_everything(self.config['seed_everything'])
+            seed_everything(self.config['seed_everything'], workers=True)
         self.before_instantiate_classes()
         self.instantiate_classes()
-        self.prepare_fit_kwargs()
-        self.before_fit()
-        self.fit()
-        self.after_fit()
+        self.prepare_run_kwargs()
+        self.before_run()
+        self.run()
+        self.after_run()
 
     def init_parser(self) -> None:
         """Method that instantiates the argument parser"""
         self.parser = LightningArgumentParser(**self.parser_kwargs)
 
     def add_core_arguments_to_parser(self) -> None:
         """Adds arguments from the core classes to the parser"""
+        self.parser.add_argument(
+            'trainer_fn', type=TrainerFn, default=TrainerFn(self.trainer_fn), help='The trainer function to run'
```

Comment:
My idea would be the following. Not tested, so probably there are mistakes.

The `LightningCLI` class would have a property `subcommands = ['fit', 'validate', 'test', 'predict']`. This property can be easily modified when extending the class. Then there would be methods like

```python
    def init_parser(self) -> None:
        self.parser = ArgumentParser(**self.parser_kwargs)
        self.parser_subcommands = self.parser.add_subcommands()
        for subcommand in self.subcommands:
            self.add_subcommand_to_parser(subcommand)

    def add_subcommand_to_parser(self, subcommand) -> None:
        parser = LightningArgumentParser()
        self.add_core_arguments_to_parser(parser)
        parser.add_method_arguments(self.trainer_class, subcommand, subcommand, skip={'model', 'datamodule', ...})
        self.add_arguments_to_parser(parser)
        method = getattr(self, subcommand)
        self.parser_subcommands.add_subcommand(subcommand, parser, help=new_function_to_get_docstring(method))
```

The result of `parse_args` will be something like `{"subcommand": "predict", "predict": {"trainer": {...}, "data": ...}}`. Based on this it can be decided what needs to be executed and with which arguments.

Running `python tool.py fit --print_config` will have the same result as now. To run fit it would be like `python tool.py fit --config config.yaml`.

Reply:
Thanks! This is definitely better. Some questions:

- Do you think we should add the core arguments to the subcommand parser (what your propose) or to the base parser?
- Can we set `fit` to be the default subcommand to use? And I would like that this is configurable, e.g. `LightningCLI(trainer_fn="test")` so test become the default subcommand
- Any details for how to implement `new_function_to_get_docstring()`?

Reply:
> Do you think we should add the core arguments to the subcommand parser (what your propose) or to the base parser? I think I prefer the later.

Yes, could be in the base parser. This will require extending in the documentation the explanation on how to use `print_config` with the subcommands. To avoid confusions the `--config` and `--print_config` options should only be available before the subcommand. This is because arguments after the subcommand behave differently. With the latest version of jsonargparse (v3.12.0) the `print_config` option is only added when `ActionConfigFile` is added, which will help in this. What needs to be explained is that `print_config` expects an optional value. This means that doing `tool.py --print_config predict` will fail because `predict` is not a valid option for `print_config`. An explicit empty option must be given e.g. `tool.py --print_config '' predict` or a valid option. This is if you want the printed config to include the defaults for the subcommand.

> Can we set fit to be the default subcommand to use? And I would like that this is configurable, e.g. LightningCLI(trainer_fn="test") so test become the default subcommand

There is no such thing as a default subcommand. This is not a good idea to implement in a general sense because arguments before and after a subcommand are interpreted in a different context. Also having a default subcommand would be confusing for the user in general argument parsing use cases. Just imagine that for `git` the default subcommand would be `commit`. Being required to give the subcommand is not much to ask and it is better to be explicit.

Reply:
> >  Do you think we should add the core arguments to the subcommand parser (what your propose) or to the base parser? I think I prefer the later.
> Yes, could be in the base parser.

With what I originally proposed the subcommand would always be first which is a bit more natural. I might be more inclined for my original proposal.

Reply:
> Any details for how to implement new_function_to_get_docstring()?

```python
import docstring_parser

def get_short_description(component) -> Optional[str]:
    try:
        docstring = docstring_parser.parse(component.__doc__)
        return docstring.short_description
    except ValueError:
        warning(f'Failed parsing docstring for {component}')
```

Reply:
> This will require extending in the documentation the explanation on how to use print_config with the subcommand

Will focus on docs after the implementation is final

> To avoid confusions the --config and --print_config options should only be available before the subcommand. 

Does this mean the `ckpt_path` (for example) cannot be configured with the config?

> `tool.py --print_config '' predict`

I don't understand why we need `''`. You are still passing the subcommand.

> Being required to give the subcommand is not much to ask and it is better to be explicit.

Okay. It's just a bit unfortunate because 90% of people will call fit, and fit does not take any subcommand arguments so having to pass it is pointless in that case.

> With what I originally proposed the subcommand would always be first which is a bit more natural.

I thought about this yesterday. Some issues:

- Adding all the core arguments to each subcommand parser seems silly.
- We would have to nest arguments like `seed_everything` with a key, otherwise we get conflicts
- The user would have to pass subcommand options as `tool.py test --test.ckpt_path` (repeating the test key)

Reply:
> > `tool.py --print_config='' predict`
> 
> Okay, I had no idea `print_config` wasn't just a boolean flag. What values would it take?

Currently it accepts `skip_null` and `comments`.

> Can this be fixed? It is very counter-intuitive

Be fixed how? Always require a value? Without the value it is not possible to customize the output.

> > This is if you want the printed config to include the defaults for the subcommand
> 
> Yes. But if `''` is not passed to `print_config` the command fails entirely

It fails saying what the problem is. Which is like any mistake you make in an argument. You just fix it and run it again.

> > Being required to give the subcommand is not much to ask and it is better to be explicit.
> 
> Okay. It's just a bit unfortunate because 90% of people will call fit, and fit does not take any subcommand arguments so having to pass it is pointless in that case.

It is not pointless. I have never seen a command line tool that supports subcommands and has a default subcommand. The point is that writing the command makes it clear what you want to execute. And it is not an issue that `fit` does not receive arguments. After the subcommand there must be at least one argument, e.g. `tool.py fit --config=config.yaml`.

> > With what I originally proposed the subcommand would always be first which is a bit more natural.
> 
> I thought about this yesterday. Some issues:
> 
>     * Adding all the core arguments to each subcommand parser seems silly.

"seems silly" is not an issue. What needs to make sense is the usage. How to implement it follows from that. If what makes sense needs to be implemented by adding all arguments to each subcommand, then this is perfectly fine.

>     * We would have to nest arguments like `seed_everything` with a key, otherwise we get conflicts

This is not true. What I proposed was that all arguments are given after the subcommand. It would be run like `tool.py fit --config=config.yaml --seed_everything=123`. There is no need to nest arguments and there will be no conflicts.

>     * The user would have to pass subcommand options as `tool.py test --test.ckpt_path` (repeating the test key)

The arguments for the function could be added without nesting. We just need a way to keep track of which are the arguments to know how to call the function.

Reply:
> Be fixed how?

Fixed so `tool.py --print_config predict` == `tool.py --print_config='' predict`

Reply:
> > Be fixed how?
> 
> Fixed so `tool.py --print_config predict` == `tool.py --print_config='' predict`

Figured out a way to fix it with an acceptable compromise. It will be valid in any location `--print_config` and `--print_config=value`. But it will not be valid to have the value separated by a space `--print_config value`.

Reply:
> > ```
> > * The user would have to pass subcommand options as `tool.py test --test.ckpt_path` (repeating the test key)
> > ```
> 
> The arguments for the function could be added without nesting. We just need a way to keep track of which are the arguments to know how to call the function.

To keep track I could modify `add_method_arguments` so that it returns a list with the arguments that were added.